gen/constant.go

```diff
@@ -20,17 +20,222 @@
 
 package gen
 
-import "github.com/thriftrw/thriftrw-go/compile"
+import (
+	"fmt"
+	"strconv"
 
-// Constant TODO
+	"github.com/thriftrw/thriftrw-go/compile"
+)
+
+// Constant generates code for `const` expressions in Thrift files.
 func Constant(g Generator, c *compile.Constant) error {
 	err := g.DeclareFromTemplate(
-		`
-		const <goCase .Name> <typeReference .Type> = nil // TODO
-		`,
+		`<if canBeConstant .Type>const<else>var<end> <goCase .Name> <typeReference .Type> = <constantValue .Value .Type>`,
 		c,
+		TemplateFunc("constantValue", ConstantValue),
+		TemplateFunc("canBeConstant", canBeConstant),
 	)
-	// TODO(abg): Implement constantValue
-	// TODO(abg): Complex types will need to use var rather than const
 	return wrapGenerateError(c.Name, err)
 }
+
+// ConstantValue generates an expression containing the given constant value of
+// the given type.
+//
+// The constant must already have been linked to the given type.
+func ConstantValue(g Generator, c compile.ConstantValue, t compile.TypeSpec) (string, error) {
+	switch v := c.(type) {
+	case compile.ConstantBool:
+		if v {
+			return "true", nil
+		}
+		return "false", nil
+	case compile.ConstantDouble:
+		return fmt.Sprint(float64(v)), nil
+	case compile.ConstantInt:
+		return fmt.Sprint(int(v)), nil
+	case compile.ConstantList:
+		l, ok := t.(*compile.ListSpec)
+		if !ok {
+			return "", fmt.Errorf(
+				"cannot cast list %v to %v", v, t.ThriftName())
```

Comment:
It looks like,

    cannot cast [1 2 3] to foo

Although this specific error won't usually happen in this part of the code because the compile step already type checked everything. It *would* be okay to panic here but I figured it's safer to fail with an actual error regardless.